s3tests_boto3/functional/test_iam.py

```diff
@@ -2052,3 +2052,179 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
+                                                VersionId=version_ids[0])
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = s3_client_iam.get_object_tagging(Bucket=bucket, Key=obj_key,
+                                                VersionId=version_ids[0])
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    eq(response['TagSet'], tags['TagSet'])
+
+    # Apply Deny PutObjectVersionTagging policy
+    deny_put_version_tag_policy = json.dumps(
+        {
+            "Version": "2012-10-17",
+            "Statement": {
+                "Effect": "Deny",
+                "Action": "s3:PutObjectVersionTagging",
+                "Resource": f"arn:aws:s3:::{bucket}/*"
+            }
+        }
+    )
+    response = client.put_user_policy(PolicyDocument=deny_put_version_tag_policy,
+                                      PolicyName='PutVersionTag', UserName=get_iam_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+    # Try put_object_tagging with version, expect AccessDenied
+    for version_id in version_ids:
+        e = assert_raises(ClientError, s3_client_iam.put_object_tagging, Bucket=bucket,
+                          Key=obj_key, VersionId=version_id, Tagging=tags)
+        status, error_code = _get_status_and_error_code(e.response)
+        eq(status, 403)
+        eq(error_code, 'AccessDenied')
+
+    # Try put_object_tagging without version, should succeed
+    response = s3_client_iam.put_object_tagging(Bucket=bucket, Key=obj_key, Tagging=tags)
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+    # Replace policy
+    allow_put_version_tag_policy = json.dumps(
+        {
+            "Version": "2012-10-17",
+            "Statement": {
+                "Effect": "Allow",
+                "Action": "s3:PutObjectVersionTagging",
+                "Resource": f"arn:aws:s3:::{bucket}/*"
+            }
+        }
+    )
+    # Set and Get object version tags
+    response = client.put_user_policy(PolicyDocument=allow_put_version_tag_policy,
+                                      PolicyName='PutVersionTag', UserName=get_iam_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = s3_client_iam.put_object_tagging(Bucket=bucket, Key=obj_key, Tagging=tags,
+                                                VersionId=version_ids[0])
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = s3_client_iam.get_object_tagging(Bucket=bucket, Key=obj_key,
+                                                VersionId=version_ids[0])
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    eq(response['TagSet'], tags['TagSet'])
+
+    # Cleanup - Delete policies
+    response = client.delete_user_policy(PolicyName='PutVersionTag',
+                                         UserName=get_iam_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
```

Comment:
cleanup of buckets and objects if not covered.

Reply:
Done